### PR TITLE
Cities needlessly drop land units in water

### PR DESCRIPTION
**Pull request by rh-github-2015**
If the tile just north of your production center happens to be coast
   and it is free and the city is garrisoned a new warrior goes swimming,
   taking an extra turn to get back onto the road network...

Please double-check for unwanted side effects. Function is at least also used to place barbarians, and I checked that and think my change can't hurt - but - four eyes see more than two.

### CODE REVIEW

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 1]
```diff
@@ -139,22 +139,31 @@ class TileMap {
...
+        // try to place at the original point (this is the most probable scenario), then move through passable tiles outwards
+        // this is done using a set, initializing it with the given position, testing whether any tile in the set is
+        // suitable (unit can move to), and if not expanding the set with all neighbors the unit could pass through, repeat.
+        // The loop is potentially done twice for land units, the first time allowing only land tiles
+        // and if this fails a second pass allows embarkment
+        val potentialCandidates = mutableSetOf<TileInfo>()
+        val isLand = unit.baseUnit.unitType.isLandUnit()
+        var allowEmbark = unit.baseUnit.unitType.isWaterUnit()
+        val isOther = !(isLand || allowEmbark)      // air and ..? to be safe, no domain check
+        while (true){
```
Comment (JackRainy) [inline core/src/com/unciv/logic/map/TileMap.kt]: Why do you need an infinite loop here, while it is enough to do only 2 iterations e.g. `for (i in 0..1)` ?

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 2]
```diff
@@ -139,22 +139,31 @@ class TileMap {
...
-                if (unitToPlaceTile != null) continue
-                // if it's not found yet, let's check their neighbours
+            while (tryCount++ < 10) {
+                unitToPlaceTile = potentialCandidates.firstOrNull { (potentialCandidates.size==1 || isLand && it.isLand || allowEmbark && it.isWater || isOther) && unit.movement.canMoveTo(it) }
+                if (unitToPlaceTile != null) break
+                // if it's not suitable, try to find another tile nearby
                 val newPotentialCandidates = mutableSetOf<TileInfo>()
                 potentialCandidates.forEach { newPotentialCandidates.addAll(getPassableNeighbours(it)) }
-                potentialCandidates = newPotentialCandidates
+                newPotentialCandidates.subtract(potentialCandidates)
```
Comment (JackRainy) [inline core/src/com/unciv/logic/map/TileMap.kt]: `subtract` - that's costly operation, why do you need that, if then goes `addAll`?
Wouldn't it better just to reassign `potentialCandidates = newPotentialCandidates` ?

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 3]
```diff
@@ -139,22 +139,31 @@ class TileMap {
...
+        // suitable (unit can move to), and if not expanding the set with all neighbors the unit could pass through, repeat.
+        // The loop is potentially done twice for land units, the first time allowing only land tiles
+        // and if this fails a second pass allows embarkment
+        val potentialCandidates = mutableSetOf<TileInfo>()
+        val isLand = unit.baseUnit.unitType.isLandUnit()
+        var allowEmbark = unit.baseUnit.unitType.isWaterUnit()
+        val isOther = !(isLand || allowEmbark)      // air and ..? to be safe, no domain check
+        while (true){
+            potentialCandidates.clear()
+            potentialCandidates.add(get(position))
```
Comment (JackRainy) [inline core/src/com/unciv/logic/map/TileMap.kt]: The initial position check can be moved out of the loop and checked before it, no need to check it twice.

**Comment by rh-github-2015** [comment]
allowEmbark *is* the 'outer loop counter' - if I start with false and fails it will reinitialize and do a second pass with true, if it starts with true there's only one pass

subtract costs, but so does add. the original went with 'all neighbors of the previous set but not necessarily the members of the old set themselves unless one is a neighbor of the other' for max ten times but had no check whether the iteration actually covered new ground. Pass two would scan a six tile hexagon with a hole in the middle, pass three would close that hole... I thought I'd go with keeping covered ground looking for new ground per pass... Come to think of it, you're right, both the old and my new code re-scan tiles already rejected. I'll optimize later when I have access to my home box... 

Though I think relying on set properties will remain a good choice. And this doesn't run too often. I briefly thought about an algorithm that would choose the 'nearest' unoccupied tile in terms of actual movement cost, which would solve the original annoyance nicely as embarkment costs 100mp, but this was quicker to do, the results pleased me well enough, and it's probably still faster too.

Wait a minute - this is the place where that 'units can move through enemy lines' quirk is coming from. So simulating actual movement **_would_** be the way to go. Definitely needs rethinking. Good catch.

**Comment by JackRainy** [comment]
> allowEmbark is the 'outer loop counter' - if I start with false and fails it will reinitialize and do a second pass with true, if it starts with true there's only one pass

I understand that. My comment is about having the fixed number of iterations is much safe and reliable (in case tomorrow somebody will break `allowEmbark` logic and the infinite loop be there).
This code brings a hidden context: it says "repeat forever" but means "repeat 1-2 times". Let's don't make the code to lie :)

> subtract costs, but so does add. 

Of course, they do, and even allocation of temporary Set does. That's my point - let's not make the situation worse. Let's keep it simple. 

> both the old and my new code re-scan tiles already rejected.

Yes, in your case it just spreads over iterations, but still happens. I can see two scenarios when `(newPotentialCandidates.size==0)` will be true after `newPotentialCandidates.subtract(potentialCandidates)` : 
1) `newPotentialCandidates` is already empty 
2) All `newPotentialCandidates` are within `potentialCandidates`

Wouldn't these checks be cheaper than `subtract` + `addAll` ?

**Comment by rh-github-2015** [comment]
I coded this assuming the basic premise - newly produced units can be placed outside the city that produced them - is OK. Reading #2100 cast doubt. The [wiki](https://civilization.fandom.com/wiki/City_%28Civ5%29) says produced units must always be placed on their city tile, so this would be pointless anyway. However, I can't believe I'd have to move the garrison anytime a new unit is about to finish... But the wiki clearly says so in the 'Gold purchasing' paragraph. That said the 10 tiles distance allowed by pre-this-pull code seemed excessive to me from the start. Does nobody have an original to check its actual behaviour? I only remember Civ4 and that had stacks.

I start to believe this function needs to differentiate between city production/purchase, free units from policies et al, barbarians and maybe cases I didn't find. City production -> max radius 0. Free unit -> civilopedia says 'near' -> max radius <= 3. Barbarians -> should additionally check for *unobserved* tiles anyway and I haven't seen where that is coded yet.

And I get now that part of your argument is about style clarity. true. For me, while true... do something... if test break is a normal construct for languages that don't have a do..until construct but yes it's not very kotlin.

**Comment by rh-github-2015** [comment]
Did I say I hate git? Branches are supposed to be independent (and in TFS they are - or they have a direct lineage and only one of them is writable). I have no idea how I could have avoided that the obsolete 'Fix: Freshly improved resources' commit from a supposedly independent branch is included here.

Anyway, [this alternative](https://github.com/yairm210/Unciv/pull/2181/commits/dfb4e3ca8ad0f321ca091c311ddf511f0b2d2285) works and looks even nicer when you buy 10k worth of scouts on an island, but it's more or less only an excercise (but it was fun to whip up except studio crashed again in that infuriating fashion where ctrl-alt-backspace is the only thing that still reponds and that's no better than pulling the plug completely, as iscsi storage gets unmounted). 

It's too over the top - and I am convinced the possibility to buy stuff and have it placed three turns away shouldn't exist in the first place. Still, the original/current master code has that ability as well. Also, I haven't checked how pathing is done elsewhere and how to synergize that with this routine - come to think of it, that's on my to-do list too, as all to often units with motv-to orders for several turns worth do not exhaust their movement points or start off in completely wrong directions...

The performance argument is of very low importance imho as the routine will not be called very often between two user interactions, or will it?

**Comment by yairm210** [comment]
I'm sorry, this is just *way too complex* to solve for the simple use-case you suggested.
I think the simplest solution would be to run "can we place this unit on a land tile" even if it's 5 tiles away. basically a 5-step BFS of "it.isLand && unit.movement.canMoveThrough(it)" 
And only if that fails, then we run "can we place this unit on a water tile", which gets rid of the it.isLand requirement
And if that turns out to have a noticable effect on efficiency (spoiler: it won't) then we can solve for that later.
For now, simplicity is key.

**Comment by rh-github-2015** [comment]
Yea I said I saw the alternative as excercise and that it's over the top.

Point is - cities shouldn't be allowed unlimited purchases therefore this shouldn't be so noticeable. Your code is a little over the top too - buy seven triplanes and the seventh has no room therefore goes to the next city within 10 tiles, possibly beyond its old 1x attack range rebase ability, and rebased for free?

See also: "if there is already a unit in the city" under Gold purchasing in the [City wiki entry](https://civilization.fandom.com/wiki/City_%28Civ5%29)
And that second-to-last paragraph in the stragegy section of the [Landsknecht wiki entry](https://civilization.fandom.com/wiki/Landsknecht_%28Civ5%29).

In other words - the usecase should be mitigated at the calling sites. If we need a routine capable of placing units in a big 10-tile radius with a reliable preference for tiles matching the unit's domain and easiest to reach from the center? I demonstrated possible implementations. But by now I think we don't. City production/Free units from policy or similar / barbarians have different requirements once unlimited city purchasing is nerfed - and we need to nerf that as we want to emulate Civ5. Until then, current code serves. I'll close and discard this, then let's wait until the city usecase is decided, then reconsider.
